### ISSUE TITLE
[NFR]: Model relation eager loading

### ISSUE THREAD

**Issue by stijn1989** [phalcon/cphalcon#15509]
I'm using Phalcon since 3.2 and I'm using this in all my projects. I started a software company and did a big project with Phalcon. Now the database contains a lot of data and they are complaining about speed. When they want to ask the overview of orders. It takes one query to fetch the orders.

`<?php
$orders = Order::find([...]);`

Each order is related to a customer. In the overview page, the name of the customer is displayed as follow:

`<?php echo $order->getCustomer()->name; ?>`

With a page with 110 orders, this only is 111 queries. I can ofcourse optimize this by sending a native query:

`SELECT * FROM orders as o INNER JOIN customers as c ON (o.customer = c.id)`

But why not make an extra parameter like follow:

`<?php
$this->belongsTo('customer', Customer::class, 'id', ['alias' => 'customer', 'eager' => true]); //fetch all field from customer
//OR
//$this->belongsTo('customer', Customer::class', 'id', [...., 'eager' => ['name]]); //fetch name field from customer only`

A sidenote to this is, when Customer has relations that need to be fetched eager, this should be done also.

That saves a lot of queries per request (because reusable = true is NOT an option for speeding the application) and less power consumption on the server :D So hooray for mother nature....

**Comment by rudiservo** [comment]
Hey @stijn1989, this is kind of a complex thing to implement, it can be done but again complex, although there my suggestion would be to use Cache like memchached to help with the load.
Also this could have a counter effect of doing more queries then it should, for example if you have a chain of eager-loading you can pull like 30 models when you actually just want one.

If you really want to optimize that view I would do something like a query with PHQL and that will solve something but not like with relations, https://docs.phalcon.io/4.0/en/db-phql.

I would be more in favour of when doing joins, or just query multiple model with PHQL,if there would have an option to keep and do the relations on the models rather then eager loading.


My idea would be something has simple has.
['select Orders.*, Customers.* from Orders where...', 'keep_relations' => true ]

Again not a very simple thing to implement but less intrusive and prone catastrophic multiplier effect, the issue would be the query would need to get the relations, check them, get the relation fields, get the data in a row, and then hydrate the model and relation when you request that current() model while you iterate.

**Comment by stijn1989** [comment]
Yes I know this is not something simple to implement. Maybe another suggestion is to set an eager statement in the find() method. That way you can choose if the model should fetch eager or not per situation. 

I'm just thinking out loudly.

**Comment by rudiservo** [comment]
> Yes I know this is not something simple to implement. Maybe another suggestion is to set an eager statement in the find() method. That way you can choose if the model should fetch eager or not per situation.
> 
> I'm just thinking out loudly.

yes that might be a way to do it has well! Good idea.
It's good to think out loud, there are no bad ideas, either they work or just need some tweeking, it's a very nice suggestion!
